README.md
# Dependencies

```
apt --no-install-recommends install novnc tigervnc-standalone-server
```

# Starting manually

X/VNC server (no security):

	Xtigervnc -geometry 1000x750 -desktop prova :100 -SecurityTypes None -localhost

X session/client:

	DISPLAY=:100 xeyes

VNC viewer (default port is 5900 + screen number)

	vncviewer localhost:6000

Export VNC port over websocket, and serve web page:

	websockify --web /usr/share/novnc/ 6080 localhost:6000

Web page for viewer:

	http://localhost:6080/vnc.html?host=localhost&port=6080&autoconnect=true

The source parsing the URL arguments is at `/usr/share/novnc/include/ui.js:36`


# `.service` files

`.service` files implement a subset of systemd's unit files. All sections and
configurations that are not supported are silently ignored.


## `[Unit]` section

Currently ignored


## `[Service]` section

These configurations are supported the same as in systemd:

* `SyslogIdentifier`
* `WorkingDirectory`
* `ExecStartPre`
* `ExecStartPost`
* `ExecStop`
* `ExecStopPost`
* `KillSignal`
* `SendSIGKILL`
* `TimeoutSec`
* `TimeoutStopSec`

These configurations are supported with some differences:

* `ExecStart`: prefixes `@`, `-`, `+` are ignored
* `KillMode`: values `control-group` and `mixed` are treated the same as `process`
* `User`, `Group`: can also take an environment variable prefixed by `$`, like
  `$SUDO_UID` or `$SUDO_GID`


## `[Webrun]` section

The `Webrun` section contains configuration directives that are specific to
webrun:

* `DisplayNumber` (default: 4): display number to use for the X VNC server
* `DisplayGeometry` (default: `800x600`): geometry of the X display where the
  application is run
* `VNCPort` (default: localhost:5900 + DisplayNumber): `host:port` to use for
  VNC communication. `host:` can be omitted and defaults to `localhost`
* `WebPort` (default: 6080): `host:port` to use for the http server / websocket
  server. `host:` can be omitted and defaults to `localhost`
* `LogFile` (default: standard error): write log output to the given file
* `WebConnectTimeout` (default: no timeout): exit if there is no http server /
  websocket connection before this amount of seconds


debian/changelog
unitd (0.~) unstable; urgency=low

  * source package automatically created by stdeb 0.8.5

 -- Enrico Zini <enrico@debian.org>  Thu, 07 Sep 2017 10:59:14 +0200


tests/test_config.py
import unittest
import textwrap
import io
from unitd.config import Parser, ParseError

def test_input(s):
    return io.StringIO(textwrap.dedent(s.lstrip("\n")))

def test_parse(s):
    return list(Parser(test_input(s)).parse())


class TestParser(unittest.TestCase):
    def test_parse(self):
        self.assertEqual(test_parse(""), [])
        self.assertEqual(test_parse("\n\n"), [])
        self.assertEqual(test_parse("\n#comment\n"), [])

        with self.assertRaises(ParseError):
            test_parse("a=1")

        l = test_parse("""
        [test]
        a=1
        a=2
        [test1]
        a=3
        """)
        self.assertEqual(l, [
            ("test", "a", "1"),
            ("test", "a", "2"),
            ("test1", "a", "3"),
        ])








tests/test_simple.py
import unittest
import tempfile
import os
from unitd import SimpleProcess, OneshotProcess
from unitd.unittest import async_test, AsyncTestCase
import unitd.config
import asyncio


class TestProcess(AsyncTestCase):
    @async_test
    def test_quits(self):
        config = unitd.config.Config()
        config.service.syslog_identifier = "test"
        config.service.exec_start.append(["/bin/true"])
        proc = SimpleProcess(config, loop=self.loop)
        yield from proc.start()
        yield from proc.terminated
        yield from proc.stop()
        self.assertEqual((yield from proc.terminated), 0)

    @async_test
    def test_oneshot(self):
        config = unitd.config.Config()
        config.service.syslog_identifier = "test"
        config.service.exec_start.append(["/bin/true"])
        proc = OneshotProcess(config, loop=self.loop)
        yield from proc.start()
        yield from proc.terminated
        yield from proc.stop()
        self.assertEqual((yield from proc.terminated), 0)

    @async_test
    def test_simple(self):
        config = unitd.config.Config()
        config.service.syslog_identifier = "test"
        config.service.exec_start.append(["/bin/sleep", "3600"])
        proc = SimpleProcess(config, loop=self.loop)
        yield from proc.start()
        yield from proc.stop()
        self.assertEqual((yield from proc.terminated), -15)

    @async_test
    def test_pre_post(self):
        config = unitd.config.Config()
        config.service.syslog_identifier = "test"
        config.service.exec_start_pre.append(["mkdir", "one"])
        config.service.exec_start_pre.append("-/bin/false")
        config.service.exec_start_pre.append(["-/bin/false"])
        config.service.exec_start_pre.append(["mkdir", "two"])
        config.service.exec_start.append(["/bin/sleep", "3600"])
        config.service.exec_start_post.append(["mkdir", "three"])
        with tempfile.TemporaryDirectory() as pathname:
            config.service.working_directory = pathname
            proc = SimpleProcess(config, loop=self.loop)

            yield from proc.start()
            self.assertTrue(os.path.isdir(os.path.join(pathname, "one")))
            self.assertTrue(os.path.isdir(os.path.join(pathname, "two")))
            self.assertTrue(os.path.isdir(os.path.join(pathname, "three")))
            yield from proc.stop()
            self.assertEqual((yield from proc.terminated), -15)


unitd/sudo.py
from contextlib import contextmanager
import os


@contextmanager
def nonroot(uid=None, gid=None):
    """
    Context manager that runs code as non-root.

    uid and gid are numeric, and if None they default to $SUDO_UID and
    $SUDO_GID from the environment
    """
    changed_gid = False
    if os.getgid() == 0:
        if gid is None:
            gid = int(os.environ["SUDO_GID"])
        os.setegid(gid)
        changed_gid = True

    changed_uid = False
    if os.getuid() == 0:
        if uid is None:
            uid = int(os.environ["SUDO_UID"])
        os.seteuid(uid)
        changed_uid = True

    yield

    if changed_uid:
        os.setuid(0)

    if changed_gid:
        os.setgid(0)


def preexec_nonroot(self, uid=None, gid=None):
    """
    Use as preexec_fn in subprocess.Popen to drop root and run with the given
    uid and gid.

    uid and gid are numeric, and if None they default to $SUDO_UID and
    $SUDO_GID from the environment

    Note that subprocess.Popen does not pass arguments to preexec_fn: use
    functools.partialmethod if you need to pass uid and gid.
    """
    if os.getgid() == 0:
        if gid is None:
            gid = int(os.environ["SUDO_GID"])
        os.setgid(int(os.environ["SUDO_GID"]))

    if os.getuid() == 0:
        if uid is None:
            uid = int(os.environ["SUDO_UID"])
        os.setuid(int(os.environ["SUDO_UID"]))



unitd/process.py
import asyncio
import signal
import sys
import os
import shlex
import logging

__ALL__ = ("SimpleProcess", "OneshotProcess")

log = logging.getLogger("unitd.process")

def create_future(loop):
    """
    Wrapper for create_future for compatibility with python 3.4
    """
    return asyncio.Future(loop=loop)
    #return loop.create_future()


class ProcessLogger:
    """
    Handle logging of stdout and stderr from a running process
    """
    def __init__(self, config, loop=None):
        self.loop = loop if loop is not None else asyncio.get_event_loop()
        self.base_log_tag = config.service.syslog_identifier
        self.log_tag = self.base_log_tag
        self.stdout_logger = None
        self.stderr_logger = None

    @asyncio.coroutine
    def _log_fd(self, prefix, fd):
        while True:
            line = yield from fd.readline()
            if not line: break
            log.debug("%s:%s", self.log_tag, line.decode('utf8').rstrip())

    def get_subprocess_kwargs(self, **kw):
        """
        Hook used when constructing arguments for
        asyncio.create_subprocess_exec
        """
        kw["stdout"] = asyncio.subprocess.PIPE
        kw["stderr"] = asyncio.subprocess.PIPE
        return kw

    def start(self, proc):
        """
        Start the logging tasks for the given process
        """
        self.log_tag = self.base_log_tag + "[{}]".format(proc.pid)
        self.stdout_logger = self.loop.create_task(self._log_fd("stdout", proc.stdout))
        self.stderr_logger = self.loop.create_task(self._log_fd("stderr", proc.stderr))

    def stop(self):
        """
        Stop the logging tasks
        """
        self.log_tag = self.base_log_tag


class Process:
    """
    Base class for processes managed by unitd
    """
    def __init__(self, config, env=None, loop=None):
        self.loop = loop if loop is not None else asyncio.get_event_loop()
        self.config = config
        self.env = env
        self.logger = ProcessLogger(config, self.loop)
        self._last_exit_code = None
        # Set to True when start succeeds, False when start fails
        self.started = create_future(self.loop)
        # Future set to the process exit code when it exits
        self.terminated = None
        # Set to True when stop succeeds
        self.stopped = create_future(self.loop)

    def _preexec(self):
        """
        Hook passed as preexec_fn to the child process
        """
        # Set user and group id
        if os.getgid() == 0 and self.config.service.group != 0:
            os.setgid(self.config.service.group)
        if os.getuid() == 0 and self.config.service.user != 0:
            os.setuid(self.config.service.user)

    def _get_subprocess_kwargs(self):
        """
        Hook used when constructing arguments for
        asyncio.create_subprocess_exec
        """
        kw = self.config.service.get_subprocess_kwargs(loop=self.loop, preexec_fn=self._preexec)
        kw = self.logger.get_subprocess_kwargs(**kw)
        if self.env is not None:
            kw["env"] = self.env
        return kw

    def _parse_cmdline(self, cmdline, flags=""):
        """
        Turn cmdline into a list of command line arguments.

        If cmdline is already a list, it is preserved as is.

        If flags is not empty, leading characters from flags in cmdline (or in
        the first element of cmdline if it is a list) are stripped, and
        returned separately.

        Returns (cmdline:list, flags:str)
        """
        if isinstance(cmdline, str):
            cmdline = shlex.split(cmdline)

        stripped = cmdline[0].lstrip(flags)
        if stripped == cmdline[0]:
            return cmdline, ""
        else:
            _flags = cmdline[0][:len(stripped)]
            cmdline[0] = stripped
            return cmdline, _flags

    @asyncio.coroutine
    def exec_wait(self, cmdline):
        """
        Run the given command line and wait for its completion.

        stdout and stderr are logged with self.logger
        """
        cmdline, flags = self._parse_cmdline(cmdline, "-")

        log.debug("%s:starting: %s", self.logger.log_tag, " ".join(shlex.quote(x) for x in cmdline))

        kw = self._get_subprocess_kwargs()
        proc = yield from asyncio.create_subprocess_exec(
            *cmdline,
            stdin=asyncio.subprocess.DEVNULL,
            **kw
        )
        self.logger.start(proc)

        self._last_exit_code = yield from proc.wait()

        log.debug("%s:exited with result %d: %s", self.logger.log_tag, proc.returncode, " ".join(shlex.quote(x) for x in cmdline))
        self.logger.stop()

        if "-" in flags:
            return True
        return proc.returncode == 0

    @asyncio.coroutine
    def _run_sync_commands(self, commands):
        """
        Run commands from the `commands` list, stopping at the first that fails
        """
        for cmd in commands:
            if not (yield from self.exec_wait(cmd)):
                return False
        return True

    @asyncio.coroutine
#    def _run_coroutine(self):
#        try:
#            yield from self._start()
#            if not self.started.cancelled():
#                yield from self._run_sync_commands(self.config.service.exec_start_post)
#                self.started.set_result(True)
#            else:
#                return
#            yield from self.proc.wait()
#            log.debug("%s:exited", self.logger.log_tag)
#        finally:
#            if self.task.cancelled():
#                log.debug("%s:cancelled", self.logger.log_tag)
#            yield from self._wait_or_terminate()

    @asyncio.coroutine
    def start(self):
        """
        Start the process.

        Return True if the process was started successfully, False if something
        failed.
        """
        if not (yield from self._run_sync_commands(self.config.service.exec_start_pre)):
            self.started.set_result(False)
            return False
        log.debug("%s:exec_start_pre succeeded", self.logger.log_tag)

        if not (yield from self._start()):
            self.started.set_result(False)
            return False
        log.debug("%s:exec_start succeeded", self.logger.log_tag)

        if not (yield from self._run_sync_commands(self.config.service.exec_start_post)):
            self.started.set_result(False)
            return False
        log.debug("%s:exec_start_post succeeded", self.logger.log_tag)

        self.started.set_result(True)
        return True

    @asyncio.coroutine
    def stop(self):
        """
        Stop the process
        """
        if self.started.done() and self.started.result():
            yield from self._run_sync_commands(self.config.service.exec_stop)
        yield from self._kill()
        yield from self._run_sync_commands(self.config.service.exec_stop_post)
        self.stopped.set_result(True)

    @asyncio.coroutine
    def _kill(self):
        """
        Kill the program
        """
        if not self.started.done():
            raise RuntimeError("{}:_kill called on a process that has not been started yet".format(self.logger.log_tag))

        try:
            if self.terminated.done():
                return

            if self.config.service.kill_mode == "none":
                return

            log.debug("%s:sending %d signal", self.logger.log_tag, self.config.service.kill_signal)
            self.proc.send_signal(self.config.service.kill_signal)

            try:
                result = yield from asyncio.wait_for(self.proc.wait(), self.config.service.timeout_stop_sec, loop=self.loop)
            except asyncio.TimeoutError:
                result = None

            if result is None and self.config.service.send_sigkill:
                log.debug("%s:sending %d signal", self.logger.log_tag, signal.SIGKILL)
                self.proc.send_signal(signal.SIGKILL)

                try:
                    result = yield from asyncio.wait_for(self.proc.wait(), self.config.service.timeout_stop_sec, loop=self.loop)
                except asyncio.TimeoutError:
                    result = None

                if result is None:
                    log.debug("%s:did not respond to SIGKILL: giving up", self.logger.log_tag)
                else:
                    log.debug("%s:exited with result code %d", self.logger.log_tag, result)
        finally:
            self.logger.stop()


class OneshotProcess(Process):
    @asyncio.coroutine
    def _start(self):
        """
        Start execution of the command, and logging of its stdout and stderr
        """
        self.terminated = create_future(self.loop)

        try:
            if not (yield from self._run_sync_commands(self.config.service.exec_start)):
                return False

            return True
        finally:
            self.terminated.set_result(self._last_exit_code)

    @asyncio.coroutine
    def _wait_or_terminate(self, wait_time=1):
        pass


class SimpleProcess(Process):
    """
    Run a command, track its execution, log its standard output and standard
    error
    """
    def __init__(self, *args, **kw):
        super().__init__(*args, **kw)
        self.proc = None
        if len(self.config.service.exec_start) != 1:
            raise RuntimeError("ExecStart should only have one entry for simple processes")
        self.cmdline, self.flags = self._parse_cmdline(self.config.service.exec_start[0], "-+@")
        self._start_confirmed = False

    def _preexec(self):
        super()._preexec()
        os.setpgrp()

    @asyncio.coroutine
    def _start(self):
        """
        Start execution of the command, and logging of its stdout and stderr
        """
        log.debug("%s:cmdline: %s", self.logger.log_tag, " ".join(shlex.quote(x) for x in self.cmdline))

        kw = self._get_subprocess_kwargs()

        self.proc = yield from asyncio.create_subprocess_exec(
            *self.cmdline, stdin=asyncio.subprocess.DEVNULL, **kw
        )
        self.logger.start(self.proc)

        task_confirm_start = self.loop.create_task(self._confirm_start())

        # Run self._confirm_start to give subclassers a way to detect when the
        # startup is done, like polling on a tcp port, or waiting for a signal
        done, pending = yield from asyncio.wait((
            task_confirm_start,
            self.proc.wait()), return_when=asyncio.FIRST_COMPLETED)

        if not self._start_confirmed:
            task_confirm_start.cancel()

        try:
            if self.proc.returncode is not None:
                log.warn("%s:failed to start (exited with code %d)", self.logger.log_tag, self.proc.returncode)
                return False
            else:
                log.debug("%s:started", self.logger.log_tag)
                return True
        finally:
            self.terminated = self.loop.create_task(self.proc.wait())

    @asyncio.coroutine
    def _confirm_start(self):
        """
        Wait until confirmation that the process has successfully started
        """
        self._start_confirmed = True


unitd/unittest.py
import unittest
import atexit
import asyncio
import functools

def async_test(f):
    @functools.wraps(f)
    def wrapper(self, *args, **kw):
        coro = asyncio.coroutine(f)
        future = coro(self, *args, **kw)
        self.loop.run_until_complete(future)
    return wrapper


class AsyncTestCase(unittest.TestCase):
    _ioloop_closed = False

    def setUp(self):
        self.loop = asyncio.get_event_loop()
        self._orig_loop_debug = self.loop.get_debug()
        self.loop.set_debug(True)
        # TODO: find a way to have a private event loop for tests
        #self.orig_event_loop = asyncio.get_event_loop()
        #self.loop = asyncio.new_event_loop()
        ##self.loop.set_debug(True)
        #asyncio.set_event_loop(None)

    def tearDown(self):
        self.loop.set_debug(self._orig_loop_debug)
        #self.loop.close()
        #asyncio.set_event_loop(self.orig_event_loop)
        pass


    @classmethod
    def _atexit(cls):
        # TODO: see python issue 23548
        if not cls._ioloop_closed:
            asyncio.get_event_loop().close()


atexit.register(AsyncTestCase._atexit)


unitd/processpool.py
import asyncio
import logging
from .signals import create_future_for_signal

log = logging.getLogger("unitd.processpool")


class ProcessPool:
    def __init__(self, loop=None):
        self.loop = loop if loop is not None else asyncio.get_event_loop()
        # Future that will get a result if the quit signal is received
        self.quit_signal = None
        # Set to False when a process fails to start
        self.success = True
        # List of processes that we manage
        self.processes = []

    def set_quit_signal(self, sig):
        log.debug("Installing handler for signal %d", sig)
        self.quit_signal = create_future_for_signal(sig)

    @asyncio.coroutine
    def start_sync(self, process):
        if not self.success:
            log.debug("A previous process failed to start, %s will not be started", process.logger.log_tag)
            return False

        if self.quit_signal.done():
            log.debug("Quit signal received, %s will not be started", process.logger.log_tag)
            return False

        log.debug("Starting process %s synchronously", process.logger.log_tag)
        res = yield from process.start()
        if not res:
            self.success = False
        self.processes.append(process)
        return self.success

    @asyncio.coroutine
    def run(self):
        try:
            if not self.success:
                return

            wait_for = [p.terminated for p in self.processes]
            if self.quit_signal:
                wait_for.append(self.quit_signal)

            done, pending = yield from asyncio.wait(
                wait_for,
                return_when=asyncio.FIRST_COMPLETED,
                loop=self.loop)
        finally:
            yield from asyncio.wait(
                [p.stop() for p in self.processes],
                return_when=asyncio.ALL_COMPLETED)


unitd/config.py
import re
import os
import pwd
import grp
import signal
from contextlib import contextmanager
import logging

log = logging.getLogger("unitd.config")

class ParseError(RuntimeError):
    def __init__(self, fd, lineno, msg):
        self.fd = fd
        self.filename = getattr(fd, "name", None)
        self.lineno = lineno
        if self.filename is None:
            msg = "line {}:{}".format(self.lineno, msg)
        else:
            msg = "{}:{}:{}".format(self.filename, self.lineno, msg)
        super().__init__(self, msg)

re_empty = re.compile(r"^\s*(?:#|$)")
re_section = re.compile(r"^\[(?P<section>[^]]+)\]\s*$")
re_assign = re.compile(r"^\s*(?P<key>\w+)\s*=\s*(?P<val>.+?)\s*$")


class HostPort:
    """
    Hold a host:port pair
    """
    def __init__(self, host="localhost", port=None):
        self.host = host
        self.port = port

    def __str__(self):
        if self.host is not None:
            return "{}:{}".format(self.host, self.port)
        else:
            return str(self.port)

    @classmethod
    def parse(cls, s):
        mo = re.match(r"^(?:(?P<host>[^:]+):)?(?P<port>\d+)$", s)
        if not mo:
            raise ValueError("invalid [host:]port: `{}`".format(s))
        if mo.group("host"):
            return cls(host=mo.group("host"), port=int(mo.group("port")))
        else:
            return cls(port=int(mo.group("port")))


class Parser:
    def __init__(self, fd):
        self.fd = fd
        self.lineno = None

    def parse_error(self, msg):
        raise ParseError(self.fd, self.lineno, msg)

    def parse(self):
        current_section = None

        for self.lineno, line in enumerate(self.fd, start=1):
            if re_empty.match(line): continue

            mo = re_section.match(line)
            if mo:
                current_section = mo.group("section")
                continue

            mo = re_assign.match(line)
            if mo:
                if current_section is None:
                    self.parse_error("key=value line found outside all sections")
                yield current_section, mo.group("key"), mo.group("val")
                continue

            self.parse_error("line not recognised as comment, [section] or assignment")

    def parse_bool(self, s):
        if s.lower() in ("yes", "true", "1"):
            return True
        elif s.lower() in ("no", "false", "0"):
            return False
        else:
            self.parse_error("invalid bool value: `{}`".format(s))

    def parse_int(self, s):
        try:
            return int(s)
        except ValueError:
            self.parse_error("invalid integer value: `{}`".format(s))

    def parse_signal(self, s):
        if re.match("^[A-Z]+^", s):
            signum = getattr(signal, s, None)
            if signum is None:
                self.parse_error("invalid signal name: `{}`".format(s))
            return signum
        else:
            try:
                return int(s)
            except ValueError:
                self.parse_error("invalid signal number: `{}`".format(s))

    def parse_delay(self, s):
        if s == "infinity":
            return None
        if s.isdigit():
            return self.parse_int(s)
        re_delay = re.compile("^(?:(?P<min>\d+)min|(?P<sec>\d+)sec)$")
        res = 0
        for delay in s.split():
            mo = re_delay.match(delay)
            if not mo:
                self.parse_error("invalid time unit: `{}`".format(s))
            val = mo.group("min")
            if val is not None:
                res += int(val) * 60
                continue
            val = mo.group("sec")
            if val is not None:
                res += int(val)
        return res

    def parse_user(self, s):
        if s.startswith('$'):
            s = os.environ[s[1:]]
        if s.isdigit():
            return int(s)
        try:
            p = pwd.getpwnam(s)
        except KeyError:
            self.parse_error("invalid user name: `{}`".format(s))
        return p.pw_uid

    def parse_group(self, s):
        if s.startswith('$'):
            s = os.environ[s[1:]]
        if s.isdigit():
            return int(s)
        try:
            g = grp.getgrnam(s)
        except KeyError:
            self.parse_error("invalid group name: `{}`".format(s))
        return g.gr_gid

    def parse_host_port(self, s):
        try:
            return HostPort.parse(s)
        except ValueError as e:
            self.parse_error(e.msg)


class Unit:
    def __init__(self):
        pass

    def from_config(self, parser, key, val):
        pass


class Service:
    def __init__(self):
        self.syslog_identifier = None
        self.working_directory = None
        self.exec_start = []
        self.exec_start_pre = []
        self.exec_start_post = []
        self.exec_stop = []
        self.exec_stop_post = []
        self.kill_mode = "control-group"
        self.kill_signal = signal.SIGTERM
        self.send_sigkill = True
        self.timeout_stop_sec = 2
        self.user = os.getuid()
        self.group = os.getgid()

    def get_subprocess_kwargs(self, **kw):
        if self.working_directory is not None:
            kw["cwd"] = self.working_directory
        return kw

    def from_config(self, parser, key, val):
        if key == "SyslogIdentifier":
            self.syslog_identifier = val
        elif key == "WorkingDirectory":
            self.working_directory = val
        elif key == "ExecStart":
            self.exec_start.append(val)
        elif key == "ExecStartPre":
            self.exec_start_pre.append(val)
        elif key == "ExecStartPost":
            self.exec_start_post.append(val)
        elif key == "ExecStop":
            self.exec_stop.append(val)
        elif key == "ExecStopPost":
            self.exec_stop_post.append(val)
        elif key == "KillMode":
            self.kill_mode = val
        elif key == "KillSignal":
            self.kill_signal = parser.parse_signal(val)
        elif key == "SendSIGKILL":
            self.send_sigkill = parser.parse_bool(val)
        elif key == "TimeoutSec":
            self.timeout_stop_sec = parser.parse_delay(val)
        elif key == "TimeoutStopSec":
            self.timeout_stop_sec = parser.parse_delay(val)
        elif key == "User":
            self.user = parser.parse_user(val)
        elif key == "Group":
            self.group = parser.parse_group(val)

    def postprocess(self):
        if self.working_directory == "~":
            try:
                p = pwd.getpwuid(self.user)
            except KeyError:
                log.warn("cannot find passwd information for user id {}".format(self.user))
                self.working_directory = None
            self.working_directory = p.pw_dir

        if os.getuid() != 0 and os.getuid != self.user:
            log.warn("runnig as non-root: ignoring request to run as user {}".format(self.user))
        if os.getuid() != 0 and os.getgid != self.group:
            log.warn("runnig as non-root: ignoring request to run as group {}".format(self.group))

    @contextmanager
    def nonroot(self):
        """
        Context manager that runs code as non-root, using the user in this
        configuration.
        """
        changed_gid = False
        if os.getgid() == 0 and self.group != 0:
            os.setegid(self.group)
            changed_gid = True

        changed_uid = False
        if os.getuid() == 0 and self.user != 0:
            os.seteuid(self.user)
            changed_uid = True

        if changed_uid:
            try:
                user_info = pwd.getpwuid(self.user)
            except KeyError:
                user_info = None

            if user_info is not None:
                old = {}
                for key in "LOGNAME", "USER", "USERNAME", "HOME":
                    old[key] = os.environ.get(key, None)

                os.environ["LOGNAME"] = user_info.pw_name
                os.environ["USER"] = user_info.pw_name
                os.environ["USERNAME"] = user_info.pw_name
                os.environ["HOME"] = user_info.pw_dir

        yield

        if changed_uid and user_info is not None:
            for k, v in old.items():
                if v is None:
                    del os.environ[k]
                else:
                    os.environ[k] = v

        if changed_uid:
            os.setuid(0)

        if changed_gid:
            os.setgid(0)


class Webrun:
    def __init__(self):
        self.display_number = 4
        self.display_geometry = "800x600"
        self._vnc_port = None
        self.web_port = HostPort(port=6080)
        self.log_file = None
        self.web_connect_timeout = None

    def from_config(self, parser, key, val):
        if key == "DisplayNumber":
            self.display_number = parser.parse_int(val)
        if key == "DisplayGeometry":
            self.display_geometry = val
        elif key == "VNCPort":
            self._vnc_port = parser.parse_host_port(val)
        elif key == "WebPort":
            self.web_port = parser.parse_host_port(val)
        elif key == "LogFile":
            self.log_file = val
        elif key == "WebConnectTimeout":
            self.web_connect_timeout = parser.parse_int(val)

    @property
    def vnc_port(self):
        if self._vnc_port is not None:
            return self._vnc_port
        return HostPort(port=self.display_number + 5900)

    @vnc_port.setter
    def vnc_port(self, value):
        self._vnc_port = value


class Config:
    def __init__(self):
        self.unit = Unit()
        self.service = Service()
        self.webrun = Webrun()

    def read_file(self, pathname):
        """
        Read a unit file from the given pathname, returning the parsed
        Config
        """
        with open(pathname, "rt") as fd:
            parser = Parser(fd)
            for section, key, val in parser.parse():
                section = section.lower()
                if section == "service":
                    self.service.from_config(parser, key, val)
                elif section == "unit":
                    self.unit.from_config(parser, key, val)
                elif section == "webrun":
                    self.webrun.from_config(parser, key, val)

        self.service.postprocess()
